server/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportShardFlushAction.java

```diff
@@ -89,6 +103,63 @@ protected void shardOperationOnPrimary(
         }));
     }
 
+    // We are here because there was a mismatch between the SplitShardCountSummary in the request
+    // and that on the primary shard node. We assume that the request is exactly 1 reshard split behind
+    // the current state.
+    @Override
+    protected Map<ShardId, ShardFlushRequest> splitRequestOnPrimary(ShardFlushRequest request) {
+        ProjectMetadata project = projectResolver.getProjectMetadata(clusterService.state());
+        final ShardId sourceShard = request.shardId();
+        IndexMetadata indexMetadata = project.getIndexSafe(request.shardId().getIndex());
+        SplitShardCountSummary shardCountSummary = SplitShardCountSummary.forIndexing(indexMetadata, sourceShard.getId());
+        Map<ShardId, ShardFlushRequest> requestsByShard = new HashMap<>();
+        requestsByShard.put(sourceShard, request);
+        // Create a request for original source shard and for each target shard.
+        // New requests that are to be handled by target shards should contain the
+        // latest ShardCountSummary.
+        int targetShardId = indexMetadata.getReshardingMetadata().getSplit().targetShard(sourceShard.id());
+        ShardId targetShard = new ShardId(request.shardId().getIndex(), targetShardId);
+        requestsByShard.put(targetShard, new ShardFlushRequest(request.getRequest(), targetShard, shardCountSummary));
```

Comment:
Do we want to skip this logic if `shardCountSummary` is up to date (the caveat above applies)?